File: research/python-lab/src/edgechain_lab/mars/rewards.py
"""MARS reward allocation."""

from __future__ import annotations

from edgechain_lab.data.schemas import MarsScore
from edgechain_lab.fl.metrics import gini


def allocate_rewards(scores: list[MarsScore], round_budget: float) -> dict[str, float]:
    """Allocate rewards proportionally to reward-eligible composite scores."""

    eligible = [score for score in scores if score.eligible_for_reward and score.has_score > 0.0]
    if not eligible:
        return {}

    total = sum(score.composite for score in eligible)
    if total == 0.0:
        return {}

    return {score.site_id: (score.composite / total) * round_budget for score in eligible}


def reward_gini(scores: list[MarsScore], round_budget: float) -> float:
    """Compute Gini coefficient for allocated rewards."""

    rewards = allocate_rewards(scores, round_budget)
    return gini(list(rewards.values()))

File: packages/mars/src/rewards.ts
import { gini } from './math';
import type { MarsScore } from './types';

export function allocateRewards(scores: MarsScore[], roundBudget: number): Record<string, number> {
  const eligible = scores.filter((score) => score.eligibleForReward && score.hasScore > 0);
  if (eligible.length === 0) {
    return {};
  }

  const total = eligible.reduce((sum, score) => sum + score.composite, 0);
  if (total === 0) {
    return {};
  }

  return Object.fromEntries(
    eligible.map((score) => [score.siteId, (score.composite / total) * roundBudget]),
  );
}

export function rewardGini(scores: MarsScore[], roundBudget: number): number {
  return gini(Object.values(allocateRewards(scores, roundBudget)));
}
